libraries/program-error/derive/src/lib.rs

```diff
@@ -14,39 +14,72 @@
 extern crate proc_macro;
 
 mod macro_impl;
+mod parser;
 
-use macro_impl::MacroType;
-use proc_macro::TokenStream;
-use syn::{parse_macro_input, ItemEnum};
+use {
+    crate::parser::SplProgramErrorArgs,
+    macro_impl::MacroType,
+    proc_macro::TokenStream,
+    syn::{parse_macro_input, ItemEnum},
+};
 
-/// Derive macro to add `Into<solana_program::program_error::ProgramError>` traits
+/// Derive macro to add `Into<solana_program::program_error::ProgramError>`
+/// trait
 #[proc_macro_derive(IntoProgramError)]
 pub fn into_program_error(input: TokenStream) -> TokenStream {
-    MacroType::IntoProgramError
-        .generate_tokens(parse_macro_input!(input as ItemEnum))
+    let ItemEnum { ident, .. } = parse_macro_input!(input as ItemEnum);
+    MacroType::IntoProgramError { ident }
+        .generate_tokens()
         .into()
 }
 
 /// Derive macro to add `solana_program::decode_error::DecodeError` trait
 #[proc_macro_derive(DecodeError)]
 pub fn decode_error(input: TokenStream) -> TokenStream {
-    MacroType::DecodeError
-        .generate_tokens(parse_macro_input!(input as ItemEnum))
-        .into()
+    let ItemEnum { ident, .. } = parse_macro_input!(input as ItemEnum);
+    MacroType::DecodeError { ident }.generate_tokens().into()
 }
 
 /// Derive macro to add `solana_program::program_error::PrintProgramError` trait
 #[proc_macro_derive(PrintProgramError)]
 pub fn print_program_error(input: TokenStream) -> TokenStream {
-    MacroType::PrintProgramError
-        .generate_tokens(parse_macro_input!(input as ItemEnum))
+    let ItemEnum {
+        ident, variants, ..
+    } = parse_macro_input!(input as ItemEnum);
+    MacroType::PrintProgramError { ident, variants }
+        .generate_tokens()
         .into()
 }
 
 /// Proc macro attribute to turn your enum into a Solana Program Error
+///
+/// Adds:
+/// - `Clone`
+/// - `Debug`
+/// - `Eq`
+/// - `PartialEq`
+/// - `thiserror::Error`
+/// - `num_derive::FromPrimitive`
+/// - `Into<solana_program::program_error::ProgramError>`
+/// - `solana_program::decode_error::DecodeError`
+/// - `solana_program::program_error::PrintProgramError`
+///
+/// Optionally, you can add `hash_error_code_start: u32` argument to create
+/// a unique `u32` _starting_ error codes from the names of the enum variants.
+/// Notes:
+/// - The _error_ variant will start at this value, and the rest will be
+/// incremented by one
+/// - The value provided is only for code readability, the actual error code
+/// will be a hash of the input string and is checked against your input
+///
+/// Syntax: `#[spl_program_error(hash_error_code_start = 1275525928)]`
+/// Hash Input: `spl_program_error:<enum name>:<variant name>`
```

Comment:
I might be missing something, but is there a reason to include the variant name? I'm thinking it would be clearer to just use the enum name.

Reply:
Yeah good point, probably not needed now that we're only hashing variant 0